File: src/scripts/scrapePrivacyTools.js
import axios from 'axios';
import * as cheerio from 'cheerio';
import dotenv from 'dotenv';
import { supabase } from '../lib/supabase.js';

// Load environment variables
dotenv.config();

const BASE_URL = 'https://awesome-privacy.xyz';

// Configure axios defaults
axios.defaults.headers.common['User-Agent'] = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36';
axios.defaults.headers.common['Accept'] = 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8';
axios.defaults.headers.common['Accept-Language'] = 'en-US,en;q=0.5';
axios.defaults.timeout = 30000;

// Add delay between requests
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

async function scrapeCategory(url, categoryName) {
  try {
    // Random delay between 2-5 seconds
    await delay(2000 + Math.random() * 3000);

    const response = await axios.get(url);
    const $ = cheerio.load(response.data);
    
    // Extract category description
    const description = $('meta[name="description"]').attr('content') || '';
    
    // Create category slug
    const slug = url.split('/').pop();
    
    console.log(`Processing category: ${categoryName}`);

    // Insert category into Supabase
    const { data: category, error: categoryError } = await supabase
      .from('tool_categories')
      .insert({
        name: categoryName,
        description: description.trim(),
        slug: slug
      })
      .select()
      .single();

    if (categoryError) {
      console.error(`Error inserting category ${categoryName}:`, categoryError);
      return;
    }

    // Find and process tools
    const tools = [];
    $('.tool-entry').each((i, elem) => {
      const name = $(elem).find('h3').text().trim();
      const description = $(elem).find('p').text().trim();
      const link = $(elem).find('a').attr('href');
      
      // Extract GitHub URL if present
      const githubUrl = link.match(/https:\/\/github.com\/[a-zA-Z0-9-]+\/[a-zA-Z0-9-_.]+/)?.[0];
      
      // Extract license info
      const licenseMatch = description.match(/\b(?:MIT|GPL|Apache|BSD|AGPL|MPL|LGPL)(?:-\d+(?:\.\d+)?)?(?:\sLicense)?\b/i);
      const license = licenseMatch ? licenseMatch[0] : null;
      
      // Detect platforms
      const platforms = [];
      const platformKeywords = {
        'Windows': ['windows', 'win'],
        'macOS': ['macos', 'mac', 'osx'],
        'Linux': ['linux', 'ubuntu', 'debian'],
        'iOS': ['ios', 'iphone', 'ipad'],
        'Android': ['android'],
        'Web': ['web', 'browser', 'online'],
        'CLI': ['cli', 'command-line', 'terminal']
      };
      
      const descLower = description.toLowerCase();
      for (const [platform, keywords] of Object.entries(platformKeywords)) {
        if (keywords.some(keyword => descLower.includes(keyword))) {
          platforms.push(platform);
        }
      }
      
      // Extract tags
      const tags = new Set();
      const tagPatterns = [
        /#\w+/,
        /\b(?:open-source|free|paid|premium|encrypted|decentralized|p2p|self-hosted)\b/i,
        /\b(?:privacy|security|anonymity|encryption)\b/i
      ];
      
      for (const pattern of tagPatterns) {
        const matches = description.match(pattern) || [];
        matches.forEach(match => {
          tags.add(match.toLowerCase().replace('#', ''));
        });
      }

      tools.push({
        category_id: category.id,
        name,
        description: description.trim(),
        url: link,
        github_url: githubUrl,
        license,
        is_open_source: Boolean(githubUrl),
        is_free: tags.has('free'),
        has_paid_plan: tags.has('paid') || tags.has('premium'),
        platforms,
        tags: Array.from(tags)
      });
    });

    // Insert tools into Supabase with delay between each
    for (const tool of tools) {
      await delay(500); // 500ms delay between tool insertions
      
      const { error: toolError } = await supabase
        .from('privacy_tools')
        .insert(tool);

      if (toolError) {
        console.error(`Error inserting tool ${tool.name}:`, toolError);
      } else {
        console.log(`Inserted tool: ${tool.name}`);
      }
    }

    console.log(`Processed ${tools.length} tools in category ${categoryName}`);

  } catch (error) {
    if (error.response) {
      console.error(`Error scraping ${url}: ${error.response.status} - ${error.response.statusText}`);
    } else {
      console.error(`Error scraping ${url}:`, error.message);
    }
  }
}

async function scrapeAll() {
  try {
    console.log('Starting scraper...');
    
    const response = await axios.get(BASE_URL);
    const $ = cheerio.load(response.data);
    
    const categoryLinks = $('.category-link');
    console.log(`Found ${categoryLinks.length} categories`);
    
    for (let i = 0; i < categoryLinks.length; i++) {
      const link = categoryLinks[i];
      const categoryName = $(link).text().trim();
      let categoryUrl = $(link).attr('href');
      
      if (!categoryUrl.startsWith('http')) {
        categoryUrl = BASE_URL + categoryUrl;
      }
      
      console.log(`Scraping category ${i + 1}/${categoryLinks.length}: ${categoryName}`);
      await scrapeCategory(categoryUrl, categoryName);
      
      // Add longer delay between categories
      await delay(5000);
    }
    
    console.log('Scraping completed successfully!');
    
  } catch (error) {
    if (error.response) {
      console.error('Error scraping main page:', error.response.status, error.response.statusText);
    } else {
      console.error('Error scraping main page:', error.message);
    }
  }
}

// Run the scraper
scrapeAll().catch(console.error);

File: src/scripts/scrape_privacy_tools.py
import requests
from bs4 import BeautifulSoup
import json
import os
from typing import Dict, List, Optional
import re

class PrivacyToolsScraper:
    def __init__(self):
        self.base_url = 'https://awesome-privacy.xyz'
        self.categories: Dict[str, dict] = {}
        self.tools: List[dict] = []

    def clean_text(self, text: str) -> str:
        """Clean and normalize text content."""
        return ' '.join(text.strip().split())

    def extract_github_url(self, url: str) -> Optional[str]:
        """Extract GitHub URL if present."""
        github_pattern = r'https://github.com/[a-zA-Z0-9-]+/[a-zA-Z0-9-_.]+'
        match = re.search(github_pattern, url)
        return match.group(0) if match else None

    def extract_license(self, text: str) -> Optional[str]:
        """Extract license information if present."""
        license_pattern = r'\b(?:MIT|GPL|Apache|BSD|AGPL|MPL|LGPL)(?:-\d+(?:\.\d+)?)?(?:\sLicense)?\b'
        match = re.search(license_pattern, text, re.IGNORECASE)
        return match.group(0) if match else None

    def detect_platforms(self, text: str) -> List[str]:
        """Detect mentioned platforms."""
        platforms = []
        platform_keywords = {
            'Windows': ['windows', 'win'],
            'macOS': ['macos', 'mac', 'osx'],
            'Linux': ['linux', 'ubuntu', 'debian'],
            'iOS': ['ios', 'iphone', 'ipad'],
            'Android': ['android'],
            'Web': ['web', 'browser', 'online'],
            'CLI': ['cli', 'command-line', 'terminal']
        }
        
        text_lower = text.lower()
        for platform, keywords in platform_keywords.items():
            if any(keyword in text_lower for keyword in keywords):
                platforms.append(platform)
        
        return platforms

    def extract_tags(self, text: str) -> List[str]:
        """Extract relevant tags from text."""
        tags = []
        tag_patterns = [
            r'#\w+',  # Hashtags
            r'\b(?:open-source|free|paid|premium|encrypted|decentralized|p2p|self-hosted)\b',
            r'\b(?:privacy|security|anonymity|encryption)\b'
        ]
        
        for pattern in tag_patterns:
            matches = re.finditer(pattern, text, re.IGNORECASE)
            tags.extend(match.group(0).lower().replace('#', '') for match in matches)
        
        return list(set(tags))

    def scrape_category(self, url: str, category_name: str) -> None:
        """Scrape tools from a category page."""
        try:
            response = requests.get(url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')
            
            # Extract category description
            description = soup.find('meta', {'name': 'description'})
            category_desc = description['content'] if description else ''
            
            # Store category
            category_slug = url.split('/')[-1]
            self.categories[category_slug] = {
                'name': category_name,
                'description': self.clean_text(category_desc),
                'slug': category_slug
            }
            
            # Find tool entries
            tool_entries = soup.find_all('div', class_='tool-entry')  # Adjust selector as needed
            
            for entry in tool_entries:
                name = entry.find('h3').text.strip()
                description = entry.find('p').text.strip()
                link = entry.find('a')['href']
                
                # Extract additional information
                github_url = self.extract_github_url(link)
                license_info = self.extract_license(description)
                platforms = self.detect_platforms(description)
                tags = self.extract_tags(description)
                
                tool = {
                    'name': name,
                    'description': self.clean_text(description),
                    'url': link,
                    'github_url': github_url,
                    'license': license_info,
                    'is_open_source': bool(github_url),
                    'is_free': 'free' in tags,
                    'has_paid_plan': any(tag in tags for tag in ['paid', 'premium']),
                    'platforms': platforms,
                    'tags': tags,
                    'category_slug': category_slug
                }
                
                self.tools.append(tool)
                
        except Exception as e:
            print(f"Error scraping {url}: {str(e)}")

    def scrape_all(self) -> None:
        """Scrape all categories and tools."""
        try:
            # Get main page
            response = requests.get(self.base_url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')
            
            # Find category links
            category_links = soup.find_all('a', class_='category-link')  # Adjust selector as needed
            
            for link in category_links:
                category_name = link.text.strip()
                category_url = link['href']
                if not category_url.startswith('http'):
                    category_url = self.base_url + category_url
                
                print(f"Scraping category: {category_name}")
                self.scrape_category(category_url, category_name)
            
        except Exception as e:
            print(f"Error scraping main page: {str(e)}")

    def save_to_json(self) -> None:
        """Save scraped data to JSON files."""
        data = {
            'categories': self.categories,
            'tools': self.tools
        }
        
        with open('privacy_tools_data.json', 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        
        print(f"Saved {len(self.categories)} categories and {len(self.tools)} tools to privacy_tools_data.json")

def main():
    scraper = PrivacyToolsScraper()
    scraper.scrape_all()
    scraper.save_to_json()

if __name__ == "__main__":
    main()